### PR TITLE
Refactor IDE open file diagnostic analysis + some perf improvements

### PR DESCRIPTION
**Pull request by mavasani**
First step towards #44408

1. Refactor IDE open file analysis so that we have a single place where we execute the analyzers for document analysis. I have preserved the existing IDE layer diagnostic caching in this PR, but it should hopefully go away in future when we move the diagnostic computation to OOP.
2. Perf improvement: Currently, we execute open file analyzers one at a time by making separate calls into CompilationWithAnalyzers. This identified overhead where we are recomputing lot of stuff with each call and missing semantic model/bound node cache hits. This PR simplifies the approach by invoking a single syntax/semantic call into CompilationWithAnalyzers with all relevant analyzers. This required adding a new public API to CompilationWithAnalyzers to return syntax/semantic diagnostics for a tree grouped by analyzers, which has been split into a separate PR https://github.com/dotnet/roslyn/pull/45347.
3. Remove code for code fix support for CompilationEndAction diagnostics - these actions are not executed in the IDE live analysis, so the code was not executing. Additionally, we have identified that we never want to run these actions in IDE live analysis due to associated perf concerns.

Future steps:
1. Move open file diagnostic analyzer execution to OOP
2. Remove IDE layer diagnostic caching, and rely on caching at OOP side
3. Investigate moving even compiler diagnostic computation to OOP - need to ensure responsiveness of refreshing compiler diagnostics on typing is not affected

### CODE REVIEW

File: src/Features/Core/Portable/Diagnostics/AnalyzerHelper.cs [hunk 1]
```diff
@@ -298,115 +298,6 @@ private static void AssertCompilation(Project project, Compilation compilation1)
             Contract.ThrowIfFalse(compilation1 == compilation2);
         }
 
-        /// <summary>
-        /// Return all local diagnostics (syntax, semantic) that belong to given document for the given StateSet (analyzer) by calculating them
```
Comment (mavasani) [inline src/Features/Core/Portable/Diagnostics/AnalyzerHelper.cs:302]: All this code moved into `DocumentAnalysisExecutor.cs`

File: src/Features/Core/Portable/Diagnostics/AnalyzerHelper.cs [hunk 2]
```diff
@@ -624,6 +515,37 @@ IEnumerable<DiagnosticData> ConvertToLocalDiagnosticsWithCompilation()
             }
         }
 
+#if DEBUG
```
Comment (mavasani) [inline src/Features/Core/Portable/Diagnostics/AnalyzerHelper.cs:518]: Existing code moved as-is to this static helper type.

File: src/Features/Core/Portable/Diagnostics/DocumentAnalysisExecutor.cs [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+
+namespace Microsoft.CodeAnalysis.Diagnostics
+{
+    /// <summary>
+    /// Executes analyzers on a document for computing local syntax/semantic diagnostics.
+    /// </summary>
+    internal sealed class DocumentAnalysisExecutor
+    {
+        private ImmutableDictionary<DiagnosticAnalyzer, ImmutableArray<Diagnostic>>? _lazySyntaxDiagnostics;
+        private ImmutableDictionary<DiagnosticAnalyzer, ImmutableArray<Diagnostic>>? _lazySemanticDiagnostics;
```
Comment (mavasani) [inline src/Features/Core/Portable/Diagnostics/DocumentAnalysisExecutor.cs:29]: Majority of the code in this file is existing code being moved here. The only new code is caching of per-analyzer syntax/semantic diagnostics in the executor, see methods `GetSyntaxDiagnosticsAsync` and `GetSemanticDiagnosticsAsync` below and the PERF notes in these methods.

File: src/Features/Core/Portable/Diagnostics/DocumentAnalysisExecutor.cs [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+                // TODO: Move this invocation to OOP
+                var treeDiagnostics = await CompilationWithAnalyzers.GetCategorizedAnalyzerSemanticDiagnosticsAsync(model, Span, cancellationToken).ConfigureAwait(false);
+                Interlocked.CompareExchange(ref _lazySemanticDiagnostics, treeDiagnostics, null);
+            }
+
+            return _lazySemanticDiagnostics.TryGetValue(analyzer, out var diagnostics) ?
+                diagnostics :
+                ImmutableArray<Diagnostic>.Empty;
+
+            async Task<TextSpan?> GetAdjustedSpanForCompilerAnalyzerAsync()
```
Comment (mavasani) [inline src/Features/Core/Portable/Diagnostics/DocumentAnalysisExecutor.cs:212]: Local functions in this method are existing code being moved here.

File: src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.Executor.cs [hunk 5]
```diff
@@ -23,39 +23,64 @@ namespace Microsoft.CodeAnalysis.Diagnostics.EngineV2
     internal partial class DiagnosticIncrementalAnalyzer
     {
         /// <summary>
-        /// Return all local diagnostics (syntax, semantic) that belong to given document for the given StateSet (analyzer) either from cache or by calculating them
+        /// Return all cached local diagnostics (syntax, semantic) that belong to given document for the given StateSet (analyzer).
```
Comment (mavasani) [inline src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.Executor.cs:26]: We split background diagnostic computation for open files into two steps:
1. Try to get cached diagnostics, while computing the set of analyzers that do not have cached diagnostics
2. Execute all the non-cached analyzers with a single invocation into CompilationWithAnalyzers.
3. Fetch computed diagnostics per-analyzer from the above invocation, and cache and raise diagnostic reported events

In near future, the diagnostic computation invocation into CompilationWithAnalyzers will be moved to OOP, which will help simplify and/or remove the IDE layer diagnostic caching in devenv process.

File: src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnostics.cs [hunk 6]
```diff
@@ -73,29 +73,8 @@ private abstract class DiagnosticGetter
             protected ImmutableArray<DiagnosticData> GetDiagnosticData()
                 => (_lazyDataBuilder != null) ? _lazyDataBuilder.ToImmutableArray() : ImmutableArray<DiagnosticData>.Empty;
 
-            protected abstract Task<ImmutableArray<DiagnosticData>> GetDiagnosticsAsync(StateSet stateSet, Project project, DocumentId? documentId, AnalysisKind kind, CancellationToken cancellationToken);
             protected abstract Task AppendDiagnosticsAsync(Project project, IEnumerable<DocumentId> documentIds, bool includeProjectNonLocalResult, CancellationToken cancellationToken);
 
-            public async Task<ImmutableArray<DiagnosticData>> GetSpecificDiagnosticsAsync(DiagnosticAnalyzer analyzer, AnalysisKind analysisKind, CancellationToken cancellationToken)
```
Comment (mavasani) [inline src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnostics.cs:79]: Code moved as-is to sub-type as it was only invoked at one place from object of sub-type. This allowed removing dead code for `GetDiagnosticsAsync` override.

File: src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnostics.cs [hunk 7]
```diff
@@ -297,42 +296,6 @@ private bool ShouldIncludeStateSet(Project project, StateSet stateSet)
 
                 return true;
             }
-
-            protected override async Task<ImmutableArray<DiagnosticData>> GetDiagnosticsAsync(StateSet stateSet, Project project, DocumentId? documentId, AnalysisKind kind, CancellationToken cancellationToken)
```
Comment (mavasani) [inline src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnostics.cs:301]: Dead code.

File: src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnosticsForSpan.cs [hunk 8]
```diff
@@ -155,171 +144,9 @@ public async Task<bool> TryGetAsync(ArrayBuilder<DiagnosticData> list, Func<stri
                 }
             }
 
-            private async Task<bool> TryGetSyntaxAndSemanticDiagnosticsAsync(StateSet stateSet, ArrayBuilder<DiagnosticData> list, CancellationToken cancellationToken)
```
Comment (mavasani) [inline src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnosticsForSpan.cs:158]: All this code is no longer required.

File: src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnosticsForSpan.cs [hunk 9]
```diff
@@ -155,171 +144,9 @@ public async Task<bool> TryGetAsync(ArrayBuilder<DiagnosticData> list, Func<stri
...
-                if (_lazyProjectResultCache.TryGetValue(analyzer, out var result))
-                {
-                    return result.GetDocumentDiagnostics(_document.Id, AnalysisKind.NonLocal);
-                }
-
-                return ImmutableArray<DiagnosticData>.Empty;
-            }
-
-#if DEBUG
-            private void VerifyDiagnostics(SemanticModel model)
```
Comment (mavasani) [inline src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnosticsForSpan.cs:248]: Moved to a different file.

File: src/Features/Core/Portable/Diagnostics/InternalDiagnosticsOptionsProvider.cs [hunk 10]
```diff
@@ -21,8 +21,6 @@ public InternalDiagnosticsOptionsProvider()
         }
 
         public ImmutableArray<IOption> Options { get; } = ImmutableArray.Create<IOption>(
-            InternalDiagnosticsOptions.CompilationEndCodeFix,
```
Comment (mavasani) [inline src/Features/Core/Portable/Diagnostics/InternalDiagnosticsOptionsProvider.cs:24]: Code path for this option was not exercised as we do not run compilation end actions in IDE live analysis, and don't want to do so either.

**Comment by CyrusNajmabadi** [comment]
@mavasani is making it so that SemanticModelAnalyzers share teh same SemanticModel with the compiler in the scope of the work you're doing?  that would be awesome if so.  doesn't have to be in the same PR, but def something i think woudl be good to consider.

**Comment by mavasani** [comment]
> is making it so that SemanticModelAnalyzers share teh same SemanticModel with the compiler in the scope of the work you're doing? that would be awesome if so. doesn't have to be in the same PR, but def something i think woudl be good to consider.

@CyrusNajmabadi No, this change doesn't change that part. Hoping to tackle it as a follow-up item

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi No, this change doesn't change that part. Hoping to tackle it as a follow-up item

Awesome!

File: src/Compilers/CSharp/Test/Semantic/Diagnostics/DiagnosticAnalyzerTests.cs [hunk 11]
```diff
@@ -3518,5 +3518,70 @@ public B() : this(1)
                     Diagnostic("ID0001", "B").WithLocation(7, 12)
                 });
         }
+
+        [Theory, CombinatorialData]
+        public async Task TestCategorizedAnalyzerDiagnosticsAsync(bool syntax, bool singleAnalyzer)
```
Comment (mavasani) [inline src/Compilers/CSharp/Test/Semantic/Diagnostics/DiagnosticAnalyzerTests.cs:3523]: @dotnet/roslyn-compiler The compiler changes are in the first 4 files in the PR. I have also separated them in a separate commit https://github.com/dotnet/roslyn/pull/45322/commits/7087a17656b00df61058f746b4ab050faeeedb3c for easier review. Thanks.

Comment (mavasani) [reply src/Compilers/CSharp/Test/Semantic/Diagnostics/DiagnosticAnalyzerTests.cs:3523]: FYI: I have actually moved all the compiler changes in this PR to a separate PR https://github.com/dotnet/roslyn/pull/45347. This PR should have no compiler changes once that PR is merged.

File: src/Compilers/Core/Portable/PublicAPI.Unshipped.txt [hunk 12]
```diff
@@ -2,6 +2,10 @@ Microsoft.CodeAnalysis.Compilation.CreateFunctionPointerTypeSymbol(Microsoft.Cod
 Microsoft.CodeAnalysis.Compilation.CreateNativeIntegerTypeSymbol(bool signed) -> Microsoft.CodeAnalysis.INamedTypeSymbol
 Microsoft.CodeAnalysis.Diagnostics.AnalyzerFileReference.AssemblyLoader.get -> Microsoft.CodeAnalysis.IAnalyzerAssemblyLoader
 Microsoft.CodeAnalysis.Diagnostics.AnalyzerFileReference.Equals(Microsoft.CodeAnalysis.Diagnostics.AnalyzerFileReference other) -> bool
+Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers.GetCategorizedAnalyzerSemanticDiagnosticsAsync(Microsoft.CodeAnalysis.SemanticModel model, Microsoft.CodeAnalysis.Text.TextSpan? filterSpan, System.Collections.Immutable.ImmutableArray<Microsoft.CodeAnalysis.Diagnostics.DiagnosticAnalyzer> analyzers, System.Threading.CancellationToken cancellationToken) -> System.Threading.Tasks.Task<System.Collections.Immutable.ImmutableDictionary<Microsoft.CodeAnalysis.Diagnostics.DiagnosticAnalyzer, System.Collections.Immutable.ImmutableArray<Microsoft.CodeAnalysis.Diagnostic>>>
```
Comment (tmat) [inline src/Compilers/Core/Portable/PublicAPI.Unshipped.txt:5]: >Analyzer [](start = 74, length = 8)

Maybe omit "Analyzer" from the name of these methods - the methods take a an array of analyzers, so it's clear its the diagnostics for those analyzers.

File: src/Features/Core/Portable/Diagnostics/DocumentAnalysisExecutor.cs [hunk 13]
```diff
@@ -0,0 +1,307 @@
...
+            //     This is critical for better analyzer execution performance.
+            //  2. Ensure that the compiler analyzer is treated specially and does not block on diagnostic computation
+            //     for rest of the analyzers. This is needed to ensure faster refresh for compiler diagnostics while typing.
+
+            RoslynDebug.Assert(_compilationWithAnalyzers != null);
+            RoslynDebug.Assert(_compilationBasedAnalyzersInAnalysisScope.Contains(analyzer));
+
+            if (isCompilerAnalyzer)
+            {
+                return await _compilationWithAnalyzers.GetAnalyzerSyntaxDiagnosticsAsync(tree, ImmutableArray.Create(analyzer), cancellationToken).ConfigureAwait(false);
```
Comment (tmat) [inline src/Features/Core/Portable/Diagnostics/DocumentAnalysisExecutor.cs:172]: >_compilationWithAnalyzers [](start = 29, length = 25)

This invocation should also move to oop, right? (todo comment)

File: src/Features/Core/Portable/Diagnostics/DocumentAnalysisExecutor.cs [hunk 14]
```diff
@@ -0,0 +1,307 @@
...
+            RoslynDebug.Assert(_compilationWithAnalyzers != null);
+
+            var span = AnalysisScope.Span;
+            if (isCompilerAnalyzer)
+            {
+#if DEBUG
+                VerifySpanBasedCompilerDiagnostics(model);
+#endif
+
+                var adjustedSpan = await GetAdjustedSpanForCompilerAnalyzerAsync().ConfigureAwait(false);
```
Comment (tmat) [inline src/Features/Core/Portable/Diagnostics/DocumentAnalysisExecutor.cs:204]: >adjustedSpan [](start = 20, length = 12)

ditto - TODO comment

File: src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.Executor.cs (hunk 5 above)
Comment (tmat) [reply src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.Executor.cs:26]: It'd be good to include this explanation in the comments.

---
In reply to: [442895865](https://github.com/dotnet/roslyn/pull/45322#discussion_r442895865) [](ancestors = 442895865)

File: src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_IncrementalAnalyzer.cs [hunk 15]
```diff
@@ -44,33 +45,58 @@ private async Task AnalyzeDocumentForKindAsync(Document document, AnalysisKind k
...
+                    var executor = new DocumentAnalysisExecutor(analysisScope, compilationWithAnalyzers, DiagnosticAnalyzerInfoCache);
+                    foreach (var stateSet in nonCachedStateSets)
+                    {
+                        var computedData = await ComputeDocumentAnalysisDataAsync(executor, stateSet, cancellationToken).ConfigureAwait(false);
+                        PersistAndRaiseDiagnosticsIfNeeded(computedData, stateSet);
+                    }
 
-                    RaiseDocumentDiagnosticsIfNeeded(document, stateSet, kind, result.OldItems, result.Items);
+                    var asyncToken = AnalyzerService.Listener.BeginAsyncOperation(nameof(AnalyzeDocumentForKindAsync));
+                    var _2 = ReportAnalyzerPerformanceAsync(document, compilationWithAnalyzers, cancellationToken).CompletesAsyncOperation(asyncToken);
```
Comment (tmat) [inline src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_IncrementalAnalyzer.cs:78]: >_2 [](start = 24, length = 2)

nit: `_ = ReportAnalyzerPerformanceAsync(...);`?

Comment (mavasani) [reply src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_IncrementalAnalyzer.cs:78]: There is already a discard variable `_` defined in scope earlier in the method.

Comment (tmat) [reply src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_IncrementalAnalyzer.cs:85]: `_ = lhs;` does not define a variable.

**Pull request merged**